### PR TITLE
rm -rf build in test runner

### PR DESCRIPTION
**Pull request by stepancheg**
... instead of each `run` test file.

The goal here is to make tests easier to write.

Other possible future improvements:
* Make `run` file optional (assume default contents to execute main)
* Make `input` and `output` optional (assume program must exit with non-zero code)

### CODE REVIEW

File: libs/test/Test/Golden.idr [hunk 1]
```diff
@@ -209,6 +209,7 @@ runTest opts testPath = forkIO $ do
   let exe = "\"" ++ exeUnderTest opts ++ cg ++ "\""
   ignore $ system $ "cd " ++ testPath ++ " && " ++
     "sh ./run " ++ exe ++ " | tr -d '\\r' > output"
+  ignore $ system $ "cd " ++ testPath ++ " && rm -rf build"
```
Comment (CodingCellist) [inline libs/test/Test/Golden.idr:212]:
```suggestion
```

**Review by CodingCellist — CHANGES_REQUESTED** [review]
I think this is a great idea! I was actually just about to implement something like this when I saw that there already was a PR  : )
However, I would suggest moving the `rm -rf build` to the `Makefile` instead of having it as part of `runTest`. The idea being to keep the number of "exec"-like calls in Idris to a minimum. What do you think?

**Comment by stepancheg** [comment]
> The idea being to keep the number of "exec"-like calls in Idris to a minimum

Sounds like a micro-optimization which won't affect total run time measurably.

We should keep `build` removal inside test, because `build` directory removal is an essential part of the test (tests may not work correctly if build directory exists in the beginning of the test according to readme). Imagine, test runner may have an option one day: run repeatedly until test crashes: this won't work if directory removal is in the `Makefile`.

Additionally if process spawn is an issue, we should fix it by performing removal calls directly without shelling out to `rm` command.

**Comment by stepancheg** [comment]
Follow-up to this diff will be moving `build` removal to before the test run (currently it is removed after the test):
* it is convenient to be able to inspect the `build` folder after the test run
* if test was killed mid run, following run might work incorrectly if `build` is removed in the end of the test

This proposed change is not in this PR to avoid independent changes in one commit.

**Comment by gallais** [comment]
The `rm -rf build` step is quite idris specific but the test library
and its runner can be used for other executables too. I'm not too keen
on having this logic built into the library.

**Comment by stepancheg** [comment]
> The rm -rf build step is quite idris specific but the test library

`Test.Golden` is very idris specific: it knows about idris executable and idris codegen. [source](https://github.com/idris-lang/Idris2/blob/3e92863e1c0230101e8106d2622bd49a3a4d1f89/libs/test/Test/Golden.idr#L345)

> can be used for other executables too

It can be viewed as a generic testing library, but practically it is only good to run Idris tests. I would argue that it `Test.Golden` should not be a part of Idris standard library.

> I'm not too keen on having this logic built into the library.

Idris standard library is heavily undertested. Partly that's because writing a test requires too much dull work. If you have ideas how to simplify test writing, please share. What I'm thinking I have commented above.

**Comment by gallais** [comment]
> it knows about idris executable and idris codegen. source

It's not perfect but that does not mean we should go all the way in the other direction.

> It can be viewed as a generic testing library, but practically it is only good to run Idris tests.

I know for a fact that people are using this to run tests on their
own projects, using their own executables rather than Idris. We need
to be mindful of their use case.

> Partly that's because writing a test requires too much dull work. If you have ideas how to simplify test writing, please share. What I'm thinking I have commented above.

We already have templates you can `cp -R` to create new cases? I have
never manually written `rm -rf build` myself.

**Comment by stepancheg** [comment]
> I know for a fact that people are using this to run tests on their
own projects

There are two possible options:

* we can move or create a copy of the library outside of `libs` folder, in `tests` folder. More changes needed anyway as I described above.
* people can grab the library and copy it into their projects if the library is removed

> We already have templates you can cp -R to create new cases? I have never manually written rm -rf build myself.

Just comparing the cost of writing a test in Idris comparing to other projects, and the cost of writing a test is quite large. But maybe it's just me.

**Comment by mattpolzin** [comment]
Originally I liked this PR simply because it takes a line of code in a bunch of run files and reduces it to a common location in the test library which felt cleaner. Then @gallais mentioned wanting to make the test library _less_ tied to the Idris core project, not more, and that also sounds smart to me. 

Maybe there's a good way to abstract what's going on with the temporary build files so the test library can centralize cleanup without being prescriptive about what the run file does. Is the idea of "test artifacts" common enough to support an optional argument that specifies a directory relative to a test that should be deleted by the runner after each test is run? If so, that might be a path forward for this PR.

**Comment by stepancheg** [comment]
We can complicate it even further with hooks and callbacks and maintain backwards compatibility forever, but does it worth it? Generic binary testing framework comparing stdout should not be a part of a language standard library anyway. But it could be a good third-party library on GitHub.

Anyway, that testing framework has Idris-specific logic hard coded in it, and `build` folder name is specific to Idris.

I can do that hook if there’s a desire to have it but implementing additional improvements will increase this complexity even further.

**Comment by mattpolzin** [comment]
I don’t think this test library is part of a standard library. It’s a matter of semantics I suppose, but it is a separate library that happens to currently be maintained as part of the core Idris 2 project. Contrib is another such library. 

It is an awkward spot for the test library to be in, perhaps. On one hand, The Idris 2 compiler needs a test framework and it needs it to suit the needs of this project, so it makes sense for this to be maintained inside the project rather than using a third party library that might fall out of maintenance or stop being what the core project needs. On the other hand, this Golden library that the Idris 2 project motivated is definitely already useful to other projects and with such a small Idris 2 ecosystem it is really nice that third party projects can use the test library instead of it only being available for compiler testing.

**Comment by stepancheg** [comment]
> so it makes sense for this to be maintained inside the project

That’s why I proposed above: move the library into test folder. And whoever may want to use it outside of Idris project, may fork it to their project or to a standalone library somewhere on GitHub.

Also golden testing is not a rocket science really. It can be done with 10 lines of shell script or so. I understand why Idris team decided to implement it in Idris (as another test case for Idris for example), but people developing their projects don’t really need to do that.

But unit testing is reasonable have as a part of the language package. And that’s what Idris could implement/expose instead.

> it is really nice that third party projects can use the test library instead of it only being available for compiler testing

Unless it slows down the development of the Idris itself. Which I argue it does.

**Comment by mattpolzin** [comment]
I think that like Guillaume I find the overhead of adding a new test for the Idris compiler quite low already - copy an existing test or a template folder, modify slightly. It definitely could be easier, but I am also someone currently benefitting from this test library to test a library of my own so I see it from that perspective as well. Golden is currently perfectly suited for third party use to test Idris library code.

**Comment by mattpolzin** [comment]
Another thing that would make tests easier to write for the compiler are some simple options that add tests based on the templates. `make test add ...` or something along those lines.

**Comment by stepancheg** [comment]
> Golden is currently perfectly suited for third party use to test Idris library code.

I would say “perfectly” is an overstatement. You need to write **four** files per each test. And additionally maintain a list of tests. And the test itself contains quite a lot of boilerplate code (module declaration, repeated imports main function). And there’s no way to test private functions.

Golden testing has it’s uses (to test UI for example), but it is not a substitution for unit testing.

> Another thing that would make tests easier to write for the compiler are some simple options that add tests based on the templates

Also test refactoring is more expensive that way. More copypaste added to the repository, more expensive is to patch it.

**Comment by mattpolzin** [comment]
> I would say “perfectly” is an overstatement.

This is a fair critique of my statement, but my sentiment was more that it is “perfectly well suited” as in I can write my library tests _as easily_ as compiler tests are written, even if that’s not the absolute best we could possibly do. 

As for unit tests — yes please! I’d love unit testing and it’s been on the wish list (not just my personal wish list) for a while, but just as golden testing does not supersede unit testing I would say unit testing does not replace a use for golden testing. And, since we don’t have a unit testing framework and I haven’t had time to write one, I am very happy to use Golden to test my library. 

> there’s no way to test private functions. 

Small aside: opinions differ on whether unit testing _should_ test private functions anyway and not being able to directly call private functions from test cases is not an unprecedented design decision for test frameworks to make intentionally.

**Comment by stepancheg** [comment]
The PR to fork `Test.Golden` #1683 and leave `Test.Golder` as is and decide its fate later.

**Comment by gallais** [comment]
> You need to write four files per each test.

You need to `cp -R` a directory, write your test file, add the test case
to a list in your runner and run the test suite to generate the right
`expected` value.

> Golden testing has it’s uses (to test UI for example), but it is not a substitution for unit testing.

This is a golden testing library, not a unit test one. So I don't understand
the point you're trying to make here.

**Comment by stepancheg** [comment]
You also need to rename idr file and patch run file to match renamed idr file. Quite a lot of work.

My point about unit testing is that if Idris had unit testing library, these would be nearly zero external users of golden test library. Because golden testing is not exactly the most common way to test things.

**Comment by gallais** [comment]
I agree that we need a unit testing library but that should be solved by
writing a unit testing library rather than making the golden testing one
extremely ad-hoc to get closer to unit testing.

**Comment by edwinb** [comment]
I agree with @gallais here, and others. What we really need is more testing libraries (unit testing, property testing) and some of these exist in some form. We also need better testing for the libraries, the golden tests are really there to check the type checker and compiler are doing what's expected of them - and they're doing an effective job of that - they're not really intended for testing libraries even if we do do that occasionally. 

So, in the end, I would rather leave the existing test libraries and framework alone, and deal with testing libraries another way.

One nice simple way is `Refl` based testing. That is:

    test_add : 2 + 2 = 4
    test_add = Refl

We should probably consider doing more of this for the basic library functions.

**Comment by stepancheg** [comment]
I'll think of how to patch golden testing without introducing refc-specific behavior.

**Pull request closed without merging**